httpbakery/agent/agent.go

```diff
@@ -0,0 +1,181 @@
+// non-nil only if c.Key is nil.
+func SetUpAuth(c *httpbakery.Client, u *url.URL, username string) error {
+	if c.Key == nil {
+		return errgo.New("cannot set-up authentication: client key not configured")
+	}
+	SetCookie(c.Jar, u, username, &c.Key.Public)
+	c.VisitWebPage = VisitWebPage(c)
+	return nil
+}
+
+// SetCookie creates a cookie in jar which is suitable for performing agent
+// logins to u.
+func SetCookie(jar http.CookieJar, u *url.URL, username string, pk *bakery.PublicKey) {
+	al := agentLogin{
+		Username:  username,
+		PublicKey: pk,
+	}
+	b, err := json.Marshal(al)
+	if err != nil {
+		// This shouldn't happen as the agentLogin type has to be marshalable.
+		panic(errgo.Notef(err, "cannot marshal cookie"))
+	}
+	v := base64.StdEncoding.EncodeToString(b)
+	jar.SetCookies(u, []*http.Cookie{{
+		Name:  "agent-login",
+		Value: v,
+	}})
+}
+
+// VisitWebPage creates a function that can be used with
+// httpbakery.Client.VisitWebPage. The function uses c to access the
+// visit URL which should previously have been configured with SetCookie.
+// If the login fails the returned error will be of type
+// *httpbakery.Error. If the response from the visitURL cannot be
+// interpretted the error will be of type *UnexpectedResponseError.
+func VisitWebPage(c *httpbakery.Client) func(u *url.URL) error {
+	return func(u *url.URL) error {
+		req, err := http.NewRequest("GET", u.String(), nil)
+		if err != nil {
+			return errgo.Notef(err, "cannot create request")
+		}
+		resp, err := c.Do(req)
+		if err != nil {
+			return errgo.Notef(err, "cannot perform request")
+		}
+		defer resp.Body.Close()
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			logger.Errorf("cannot read response body: %s", err)
+			b = []byte{}
+		}
+		mt, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
+		if err != nil {
+			logger.Warningf("cannot parse response content type: %s", err)
+			mt = ""
+		}
+		if mt == "application/json" {
+			if resp.StatusCode == http.StatusOK {
+				var ar agentResponse
+				err := json.Unmarshal(b, &ar)
+				if err == nil && ar.AgentLogin {
+					return nil
+				}
+				if err != nil {
+					logger.Warningf("cannot unmarshal response: %s", err)
+				}
+			} else {
+				var herr httpbakery.Error
+				err := json.Unmarshal(b, &herr)
+				if err == nil && herr.Message != "" {
+					return &herr
+				}
+				if err != nil {
+					logger.Warningf("cannot unmarshal error response: %s", err)
+				}
+			}
+		}
+		uerr := (*UnexpectedResponseError)(resp)
+		uerr.Body = ioutil.NopCloser(bytes.NewReader(b))
+		return uerr
+	}
+}
+
+// UnexpectedResponseError is the error returned when a response is
+// received that cannot be interpretted.
+type UnexpectedResponseError http.Response
+
+func (u *UnexpectedResponseError) Error() string {
+	return fmt.Sprintf(
+		"unexpected response to non-interactive web page visit %s (content type %s)",
+		u.Request.URL.String(),
+		u.Header.Get("Content-Type"))
+}
+
+/*
+PROTOCOL
```

Comment:
Perhaps put this comment near the start of the file, just before type agentLogin, perhaps?